### ISSUE TITLE
Add Config Management to UI

### ISSUE THREAD

**Issue by c0fe** [containous/traefik#2481]
I propose a new feature, the ability to manage the configuration through the web UI, this would allow for easier configuration of how you want to set the various features of Traefik.

**Comment by ldez** [comment]
Every time someone does not respect the issue's template, a little bunny dies.

<details>

![southpark - estrella kill rabbit](https://user-images.githubusercontent.com/5674651/27143914-3cea7896-5130-11e7-81f8-02d8ac2da7a7.gif)
![southpark - estrella kill rabbit2](https://user-images.githubusercontent.com/5674651/27143924-49104c72-5130-11e7-82ae-ccedbd5dfa1c.gif)

</details>
<br>

Please edit your issue description to respect the [issue template](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE.md) as much as possible.

**Comment by c0fe** [comment]
@ldez The issue template is written for bugs, I fail to see how any of those except for the first section as to how it has to relate to a feature request. Why does a feature request require TOML configuration files, debug logs, what I was trying to achieve?

**Comment by papagei9** [comment]
+1

**Comment by toxik-io** [comment]
Ahem....

# feature
"I propose a new feature"
# what did you do
Doesn't apply to features, but used traefik...
# what did you expect
I'd love it if it had a web config
# debug info / version / config
The master branch's default settings.

Not every issue template is perfect. Don't shy people away from contributing with a harsh "wontfix"

**Comment by mr-manuel** [comment]
Are there any updates for a web management backend UI for changing settings/adding services and so on?

**Comment by toxic0berliner** [comment]
I would also realy appreciate something like this !
will even settle for some text editor in browser that would allow me to change traefik's config, especially dynamic config.
Possibly with some autocomplete and validations for the syntax...

It's a strange world where automation of my reverse-proxy is so easy with docker provider but at the same time I still need to use vim to edit the core config...

**Comment by endigma** [comment]
Maybe something along the lines of nginx proxy manager but for manipulating traefik static/dynamic providers

**Comment by gbakeman** [comment]
[Related discussion at the Community](https://community.traefik.io/t/feature-request-gui-to-configure-routers-and-service-or-edit-config-files/9073/2).

**Comment by mpl** [comment]
@c0fe @toxic0berliner 

As far as we understand, you would like the web UI to be able to modify some dynamic configuration part, that initially came from a provider. 
e.g. there was a label on your docker container that set a routing rule (traefik.http.routers.my-container.rule=Host(`example.com`)), which led to the creation of a router in traefik. And now you want to change that routing rule through the web UI.

But here's the problem with that: even if it was possible for the web UI to modify the final/internal merged configuration that is used as the basis for creating all the elements (routers, services, etc), then what happens when the docker infra sends an event (which will inevitably happen) and the docker provider (which has no knowledge that the web UI made that change) sends the configuration once more and "reverts" that router rule? Then the change that you made through the webUI has been reverted, and you have to do it again.

In other words, yes, we could maybe make the web UI as basically another dynamic configuration provider, but then it would be in competition with the other providers, which means it would not be viable to configure through it elements which stem from another provider. And I don't think that's what you're asking for.

So unless we missed something, we think that this feature either is impossible as intended, or would be useless if done how we can conceive it. Which is why we intend to close it, unless there's an objection.

**Comment by gbakeman** [comment]
As someone who's recently taken an interest, I want to take a stab at defending this feature request.

I think there's an underlying desire from the community of improving the UI for us as admins, and anything that can simplify the process of setting up our infrasctructure would be welcome. I come from using Portainer to manage my Docker containers, and the lack of integration starts to show when we're tediously tweaking labels to get a container talking to Traefik. Portainer doesn't have any awareness of Traefik, and it seems unlikely that it will in the near future unless someone decides to start supporting it.

However, Traefik _does_ have some level of awareness of the Docker systems which I think make it a prime candidate for this feature. Traefik is great in that it gives us several options and mediums for configuring our services, and as the Dashboard UI is already quite pretty and user-friendly, then I think it's positioned well to be taken to the next level of usefulness.

Asking questions like how do we handle conflicting information is important, but not insurmountable. As a suggestion, the web UI would exist as its own configuration provider, having the final say in how a service/entrypoint/whatever operates. Maybe we forget some extraneous labels on a container, but those would then simply be overridden by the user's config in the web UI. Maybe I'm misunderstanding the core concern here, but I think conflicts are workable. Any configuration set in the web UI IMO should not be modifiable by another provider.

**Comment by dncpax** [comment]
Well as someone who installed traefik because it had a gui and was completely surprised to see it's a readonly thingy, I would say please make it editable. The reasoning above strikes me as a bit condescending... let's not allow editing lest the users shoot themselves on the foot... of course, this is an open source project, so I'm super grateful I get to use it. But... it would be one thing if the team said there aren't enough resources for this issue, or it's just not something we wish to implement, not our philosophy, etc. Also, people use traefik for other things other than docker. You could simply set docker dynamic stuff non-editable. You see, we are going crazy with config files... how many more smart syntaxes do we need to learn?

**Comment by endigma** [comment]
Traefik supports http source for dynamic configuration, I've been evaluating options for creating a "config manager" that would operate similarly to https://nginxproxymanager.com/ and run a server that produces a dynamic config for traefik to consume.

**Comment by dncpax** [comment]
Thanks for replying. I'm sorry if I vented a bit... Traefik is an awesome product and we're lucky to get to use it.

**Comment by toxic0berliner** [comment]
@mpl I think you missed the point, I'm not asking to make impossible or useless things to edit in traefik some config that was first applied because discovered in docker provider.
But I do have a dynamic configuration folder with some files in it, applying middlewares to some entrypoints for example.
I see no argument why anything @file shouldn't be editable through the web GUI. Even static config to add an entrypoint or change it's port, I do believe it makes sense to make it editable, even if it triggers a restart of traefik and even if it messes up the comments and whatever non-structural text old-school admins type into their config files.

My bet is that once all file config options are available in the GUI, you'll see someone package traefik with a sensible default config if it ain't already the case and all tutorials and guide we'll find on the web will simply direct users to click through the GUI, to define their acme settings, to change their port numbers, to perform SSO using forward-auth or whatever...

Having used traefik for a while I do understand that one does not edit the config so often and it can feel useless as once it's setup one tends to rely on some kind of provider instead of editng text files every 2 hours...
But still, I do believe there is room to offer a way to manage what we currently do in config files inside the GUI instead. And I do not see anything preventing us from doing so except for the time needed to translate the config file concepts into the gui and actually re-writing the config file and potentially restarting.

**Comment by FabianTe** [comment]
I'd like to add one user scenario (mine): I use Traefik a lot for routing customer-specific domains onto the appropriate server in my internal network. There are both docker containers and "static" servers with some legacy software on them. Therefore I have a mix of services automatically [picked up from container labels](https://doc.traefik.io/traefik/providers/docker/) and manually configured [dynamic config files](https://doc.traefik.io/traefik/providers/file/). Additionally, similar to what @toxic0berliner mentioned, I define essential middlewares in config files which operators deploying containers should use.

I have used Consul [KV for dynamics configs](https://doc.traefik.io/traefik/routing/providers/kv/) and generally found it simpler to expose the KV UI of Consul itself so that operators are able to configure routes without the need to connect to any server to edit a file. This approach is fine and worth a try IMO if you need *any* solution that allows you to delegate route managment to someone else (a not-so-technically enabled supporter for example).

My nitpick with using the Consul KV (which applies to Redis too) is that configs are annoying to work with because every config property must be its own KV pair. I'd be happy with using a simple KV if there was an option to read KV pairs as whole configs, similar to how file configs are read when using `providers.file.directory`. That would greatly improve the UX for said supporters.

An alternative idea (that has brought me here) is to have a simpler, even lesser technical UI that allows users to simply pick a service, maybe a middleware and then define the routing rule, done. This would greatly support scenarios where I (the technical operator of a cloud environment) would be able to delegate said creation of routers to support users.

In contrast, there may be alternative ways / options to implement this use case. Hope this is sensible food for thought:

- Using the [HTTP provider](https://doc.traefik.io/traefik/providers/http/). A simple config UI could be implemented as standalone service which exposes configs via HTTP to Traefik. No need to change Traefik itself.
- Having a code editor with schema validation for Traefik configs. Such an editor could be embeded into the Traefik UI or your own service. Maybe a VSCode extension is an option.

**Comment by NextBlaubeere** [comment]
I would love to use traefik. However, the lack of ability to configure it via webinterface for quick and dirty changes is keeping me off of it.

**Comment by bp2008** [comment]
3 months ago, I saw a screenshot of traefik's web UI and thought:

> These guys get it!  Their goal was easy deployment so they built an awesome web based configuration system!  Here it is in this picture!

![image](https://github.com/traefik/traefik/assets/5639911/1aac8aff-2a27-4c25-951d-4141a5e04016)

...

Of course as I got started, I quickly realized the web interface is read-only! :(

It seems all of traefik's automation is dependent on complex docker setups or kubernetes, none of which I have or want.  I was intending to do 100% manual setup of ports, hostnames, routes, etc.  For my purposes, traefik would be just as dependent on manually editing configuration files as nginx which I was trying to escape.

What I ended up doing was [build my own](https://github.com/bp2008/WebProxy/wiki/Admin-Console-Tour) reverse proxy application from scratch.  One that deploys easily on a fresh machine with minimal command line work and _zero_ editing of configuration files.  I've been wanting this tool on my belt for years.  Now I have it.  Its feature set will never be in the same league as traefik, and I don't know that I would ever recommend using it for high value production workloads.  But it is out there if anyone wants to try it (who hates configuration files as much as I do).

**Comment by bp2008** [comment]
> Not sure many will want to switch to a homemade reverse proxy entirely bypassing traefik

Oh, certainly not.  It is basically a guarantee that traefik is more performant, better tested, and more compatible with some of the weird HTTP servers out there.

> there is no good reason that I know of to keep read-only any settings that's stored in a traefik config file

I can think of one.  A write-enabled configuration UI accessible via HTTP is an attack vector that previously did not exist, and must be taken seriously from a security standpoint.  Even with credentials and an IP whitelist or restrictive firewall, attacks such as [CSRF](https://owasp.org/www-community/attacks/csrf) are possible and should be accounted for.

**Comment by stavros-k** [comment]
Then you will also need a role based access control with granular control.

Then also you need auditing, versioning, easy rollback. Cause you don't want to misclick and take prod down.

Traefik is designed for containers mostly. Not for someone that has few services on his homelab.

Sure ui editing for those people is nice. But I don't expect anything like that anytime soon.

**Comment by toxic0berliner** [comment]
The dashboard is disabled by default, I seem to remember it not being recommended for production also. It can't be that for security reasons you're left with your terminal and ssh to update the config. 
That said I do hope any UI config would go through to writing the associated config file and making these files read only might be a safe way to "disable" the web UI write mode. Could also be an option to keep the dashboard read only in the config file itself.
Without further design or technical proposal I disagree that security is a valid reason to decide to not even look at this enhancement.
I do understand it's a big one or at least not a small one. I do understand the devs might not care nor have the time or skills. But there are safe ways to build this.

**Comment by toxic0berliner** [comment]
> Then you will also need a role based access control with granular control.
> 
> Then also you need auditing, versioning, easy rollback. Cause you don't want to misclick and take prod down.
> 
> Traefik is designed for containers mostly. Not for someone that has few services on his homelab.
> 
> Sure ui editing for those people is nice. But I don't expect anything like that anytime soon.
> 
> IasC is getting more and more popular and that's a good thing

Noone here talked about rbac. Sure you could imagine this as a next step but this is already true for the read only dashboard . 
Paranoid users already fear the vast amount of information being exposed by the API and it can sometimes correctly be seen an empowering tool for potential attackers, didn't stop the project to still build the dashboard and not start right away with rbac to limit one's view...

**Comment by stavros-k** [comment]
> Noone here talked about rbac. Sure you could imagine this as a next step but this is already true for the read only dashboard . Paranoid users already fear the vast amount of information being exposed by the API and it can sometimes correctly be seen an empowering tool for potential attackers, didn't stop the project to still build the dashboard and not start right away with rbac to limit one's view...

The moment you open "editing" on the webui, you have to have a way to restrict that to specific people.
You don't want your friend that came to your home lan suddenly "play" with that edit UI or an employ that should only be able to view, now to be able to edit.
Currently that can be easily done with service account in k8s, or required approvers on the versioning system.

> And "meant for containers" doesn't mean you don't need traefik config. The argument here is the config you do need from traefik should be editable without a terminal. I don't believe it's possible and if it is surely not widely used that the entry points themselves are being defined by another dynamic provider... File provider should I believe be gui-editable

My k8s cluster do not even have ssh access. (Talos linux) traefik is fully configurable with CRDs.

Yes, on docker you need couple of config files (sadly) as some very few fields are not configurable via labels.
But most of them are.

---

To be clear, I don't say "Don't do it". Heck I have even upvoted it.
I just say that if it ever comes, it will be when all or most of the above things I've mentioned are there.
Cause traefik is a company and targets businesses. Sure its opensource and benefits from contriburtors.
But I don't think this feature will be greenlighted until it checks some boxes for a business user.


I think the best move would be to build an editable UI as a separate project that exposes an API and traefik can consume that.
As @fabiante already mentioned. 

Luckily with Go, it should be easy to import all the validation and structures and build on top of that.

**Comment by Hutch79** [comment]
Hey everyone

Just landed here after looking (again) for an alternative to NginxProxyManager (NPM).
I would love to use Traefik. But I despise editing configs manually. This was the sole reason why I used and currently still using NPM despite it's many problems.

I read most of the comments above, and I see the Traefik is mostly made for a more complex infrastructure which runs some sort of container orchestrator like K8s. 
Despite that, I'd argue that a configuration interface would bring some value to the product overall.
Not nessecarely to advanced Admins which are already using Traefik. They are (hopefully) happy to use it in its current state. But I think it would make the entry to use Traefik much easier for more novel use cases like a home lab.

I'd say that the Web Configurator would not need to provide every single feature which is available in Traefik. I personally would also be willing to set up everything in a config file. The most important thing for me in the WebUI would be adding and deleting routs. So basically "Take this subdomain and point it there". Basically the functionality NPM provides.

I've also seen the concern about security.
I would suggest that the configuration functionality is disabled by default.
And if enabled, i see two options to secure it.
1. (the ugly) Create a second Frontend purely for configuration. This could be secured by a third party auth proxy like Authentik.
2. (the prevered) Implement a basic authentication system. The current dashboard is still free to access as before, but the edit stuff is hidden behind a password which can be determent in an env var. No user management, no Role based access control, no config versioning. Those features are nice, but not necessary. If you fucked up, you fucked up.

About the concern of overwriting docker flags.
I guess the simplest solution would be to deny such edits.
The only thing the Web Editor needs to do is to edit the config file which is used by the config provider.

I hope that this feature will be added in the future. It's the only feature I'm really missing in this otherwise amazing project.

**Comment by GuyFromMars365** [comment]
Considering it's been 7 years since this request has been opened, I would,'t ever expect implementation, though I'm 100% in favor of GUI config option. Would massively reduce user complication. I would absolutely pay for a premium option for GUO-based config.

**Comment by unf0rg0tt3n** [comment]
It's now 2026 and still there isn't a dashboard that can be edited

**Comment by fabiante** [comment]
Hi @unf0rg0tt3n - I am not a contributor to Traefik, but i think it is safe to say that this won't become a feature.

I think this statement from above is very sensible:

> I think the best move would be to build an editable UI as a separate project that exposes an API and traefik can consume that.

While on the surface such a UI seems an "easy to add" feature, the devil is very much in the details. With the whole ecosystem around Traefik, it is a cleaner approach to have a separate web UI for this.

My personal use case for this does not exist anymore, but if i would still have it, i would explore whipping something simple using AI generation. Alternatively, this seems like a great open-source personal project to maintain 😁 😉

**Comment by unf0rg0tt3n** [comment]
> Hi [@unf0rg0tt3n](https://github.com/unf0rg0tt3n) - I am not a contributor to Traefik, but i think it is safe to say that this won't become a feature.
> 
> I think this statement from above is very sensible:
> 
> > I think the best move would be to build an editable UI as a separate project that exposes an API and traefik can consume that.
> 
> While on the surface such a UI seems an "easy to add" feature, the devil is very much in the details. With the whole ecosystem around Traefik, it is a cleaner approach to have a separate web UI for this.
> 
> My personal use case for this does not exist anymore, but if i would still have it, i would explore whipping something simple using AI generation. Alternatively, this seems like a great open-source personal project to maintain 😁 😉

There is a dashboard built by another person (https://github.com/chr0nzz/traefik-manager). I haven't tried it yet because wanted to explore the official way. I just ask gemini-cli to edit the config. And run it by me.
